### ISSUE TITLE
v3.3.0-rc.0 restore from same snapshot, cannot add a new member

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#9109]
From https://github.com/coreos/etcd/issues/9094#issuecomment-355571709

1. save snapshot
2. restore 3-node cluster from same snapshot
3. add a new member (fails)

We don't have any test coverage around this...

Likely related to https://github.com/coreos/etcd/issues/9096.

/cc @lyddragon

**Comment by lyddragon** [comment]
how to do normal restore etcd cluster?

**Comment by gyuho** [comment]
@lyddragon Can't reproduce. What was the error message from 

> sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node4 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380,node4=http://192.168.0.80:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.80:2380 --data-dir /data/etcd/intranet-test.data

> start failed

**Comment by lyddragon** [comment]
1、get keys:

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k1
> 1
> k2
> 2
> k3
> 3
> k4
> 4

2、backup on node1，and scp s.db to node2、node3

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot save /tmp/s.db
> $ scp /tmp/s.db node2:/tmp
> $ scp /tmp/s.db node3:/tmp

3、on node1、node2、node3

> $ systemctl stop etcd
> $ rm -rf /data/etcd/*

4、on node1、node2、node3，restore and start etcd service

> node1：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node1 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.81:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 
> node2：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node2 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.82:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 
> node3：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node3 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.83:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 

5、add member node4

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl member add node4 --peer-urls='http://192.168.0.80:2380'


6、start node4

> $ cat /usr/local/etcd/conf/etcd.conf
> ......
> initial-cluster-state: existing
> ......
> $ systemctl start etcd

7、on node4, get keys, no keys

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'

8、on node1, and add keys

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k1
> 1
> k2
> 2
> k3
> 3
> k4
> 4

> 
$ ETCDCTL_API=3 /usr/local/etcd/etcdctl put k5 5

9、on node4， get keys，no k1-k4

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k5
> 5

**Comment by lyddragon** [comment]
1、get keys:

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k1
> 1
> k2
> 2
> k3
> 3
> k4
> 4

2、backup on node1，and scp s.db to node2、node3

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot save /tmp/s.db
> $ scp /tmp/s.db node2:/tmp
> $ scp /tmp/s.db node3:/tmp

3、on node1、node2、node3

> $ systemctl stop etcd
> $ rm -rf /data/etcd/*

4、on node1、node2、node3，restore and start etcd service

> node1：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node1 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.81:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 
> node2：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node2 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.82:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 
> node3：
> $ sudo -u etcd ETCDCTL_API=3 /usr/local/etcd/etcdctl snapshot restore /tmp/s.db --name node3 --initial-cluster node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380 --initial-cluster-token wpbch1bi7yebkdWWfoemlqxyjbwrqt --initial-advertise-peer-urls=http://192.168.0.83:2380 --data-dir /data/etcd/intranet-test.data
> $ systemctl start etcd
> 

5、add member node4

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl member add node4 --peer-urls='http://192.168.0.80:2380'


6、start node4

> $ cat /usr/local/etcd/conf/etcd.conf
> ......
> initial-cluster-state: existing
> ......
> $ systemctl start etcd

7、on node4, get keys, no keys

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'

8、on node1, and add keys

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k1
> 1
> k2
> 2
> k3
> 3
> k4
> 4

> 
$ ETCDCTL_API=3 /usr/local/etcd/etcdctl put k5 5

9、on node4， get keys，no k1-k4

> $ ETCDCTL_API=3 /usr/local/etcd/etcdctl get 0 'z'
> k5
> 5

**Comment by gyuho** [comment]
@lyddragon What is your node4 config after running `member add` on node4?

**Comment by lyddragon** [comment]
name: "node4"
data-dir: /data/etcd/intranet-test.data
wal-dir: /data/etcd/intranet-test.wal.data
snapshot-count: 10000
heartbeat-interval: 100
election-timeout: 1000
quota-backend-bytes: 8589934592
listen-peer-urls: http://192.168.0.80:2380
listen-client-urls: http://192.168.0.80:2379,http://127.0.0.1:2379
max-snapshots: 5
max-wals: 5
cors:
initial-advertise-peer-urls: http://192.168.0.80:2380
advertise-client-urls: http://192.168.0.80:2379
discovery:
discovery-fallback: "proxy"
discovery-proxy:
discovery-srv:
initial-cluster: node1=http://192.168.0.81:2380,node2=http://192.168.0.82:2380,node3=http://192.168.0.83:2380,node4=http://192.168.0.80:2380
initial-cluster-token: "wpbch1bi7yebkdWWfoemlqxyjbwrqt"
initial-cluster-state: existing
strict-reconfig-check: false
enable-v2: false
enable-pprof: true
proxy: "off"
proxy-failure-wait: 5000
proxy-refresh-interval: 30000
proxy-dial-timeout: 1000
proxy-write-timeout: 5000
proxy-read-timeout: 0
client-transport-security:
ca-file:
cert-file:
key-file:
client-cert-auth: false
trusted-ca-file:
auto-tls: false
peer-transport-security:
ca-file:
cert-file:
key-file:
peer-client-cert-auth: false
trusted-ca-file:
auto-tls: false
debug: false
log-package-levels: etcdmain=DEBUG,etcdserver=DEBUG
log-output: default
force-new-cluster: false

**Comment by gyuho** [comment]
Did you delete data dir before start? Did you restore snapshot on node4?

**Comment by lyddragon** [comment]
node4 is new member.

**Comment by gyuho** [comment]
Did you delete data dir before starting the node4?
And can you reproduce locally?
Works fine in my side...

**Comment by lyddragon** [comment]
yes，first delete data dir.

**Comment by lyddragon** [comment]
I had tried more time.

**Comment by gyuho** [comment]
This is exactly what you are doing?

Try this yourself.
This works as expected.

If not, there's something wrong.

```bash
##################################
ETCD_VER=v3.3.0-rc.0

# choose either URL
GOOGLE_URL=https://storage.googleapis.com/etcd
GITHUB_URL=https://github.com/coreos/etcd/releases/download
DOWNLOAD_URL=${GOOGLE_URL}

rm -f /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz
rm -rf /tmp/etcd-download-test && mkdir -p /tmp/etcd-download-test

curl -L ${DOWNLOAD_URL}/${ETCD_VER}/etcd-${ETCD_VER}-linux-amd64.tar.gz -o /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz
tar xzvf /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz -C /tmp/etcd-download-test --strip-components=1
rm -f /tmp/etcd-${ETCD_VER}-linux-amd64.tar.gz

/tmp/etcd-download-test/etcd --version

/tmp/etcd-download-test/etcd
##################################

##################################
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl put foo1 bar1
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl put foo2 bar2
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl put foo3 bar3
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl get "" --from-key
##################################

##################################
rm -f /tmp/a.db

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  snapshot save /tmp/a.db

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  snapshot status /tmp/a.db --write-out=table
##################################

##################################
rm -rf /tmp/etcd-data-1.new
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  snapshot restore /tmp/a.db \
  --data-dir /tmp/etcd-data-1.new \
  --name s1 \
  --initial-advertise-peer-urls http://localhost:12380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn

rm -rf /tmp/etcd-data-2.new
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  snapshot restore /tmp/a.db \
  --data-dir /tmp/etcd-data-2.new \
  --name s2 \
  --initial-advertise-peer-urls http://localhost:22380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn

rm -rf /tmp/etcd-data-3.new
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  snapshot restore /tmp/a.db \
  --data-dir /tmp/etcd-data-3.new \
  --name s3 \
  --initial-advertise-peer-urls http://localhost:32380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn
##################################

##################################
/tmp/etcd-download-test/etcd \
  --name s1 \
  --data-dir /tmp/etcd-data-1.new \
  --listen-client-urls http://localhost:12379 \
  --advertise-client-urls http://localhost:12379 \
  --listen-peer-urls http://localhost:12380 \
  --initial-advertise-peer-urls http://localhost:12380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn \
  --initial-cluster-state existing

/tmp/etcd-download-test/etcd \
  --name s2 \
  --data-dir /tmp/etcd-data-2.new \
  --listen-client-urls http://localhost:22379 \
  --advertise-client-urls http://localhost:22379 \
  --listen-peer-urls http://localhost:22380 \
  --initial-advertise-peer-urls http://localhost:22380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn \
  --initial-cluster-state existing

/tmp/etcd-download-test/etcd \
  --name s3 \
  --data-dir /tmp/etcd-data-3.new \
  --listen-client-urls http://localhost:32379 \
  --advertise-client-urls http://localhost:32379 \
  --listen-peer-urls http://localhost:32380 \
  --initial-advertise-peer-urls http://localhost:32380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380 \
  --initial-cluster-token tkn \
  --initial-cluster-state existing
##################################

##################################
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379 \
  endpoint health

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379 \
  member list --write-out=table

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379 \
  endpoint status --write-out=table
##################################

##################################
# member add
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379 \
  member add s4 --peer-urls=http://localhost:42380

rm -rf /tmp/etcd-data-4.new
/tmp/etcd-download-test/etcd \
  --name s4 \
  --data-dir /tmp/etcd-data-4.new \
  --listen-client-urls http://localhost:42379 \
  --advertise-client-urls http://localhost:42379 \
  --listen-peer-urls http://localhost:42380 \
  --initial-advertise-peer-urls http://localhost:42380 \
  --initial-cluster s1=http://localhost:12380,s2=http://localhost:22380,s3=http://localhost:32380,s4=http://localhost:42380 \
  --initial-cluster-token tkn \
  --initial-cluster-state existing
##################################

##################################
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:42379 \
  put foo4 bar4

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379 \
  get "" --from-key

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:22379 \
  get "" --from-key

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:32379 \
  get "" --from-key

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:42379 \
  get "" --from-key
##################################

##################################
ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379,http://localhost:42379 \
  endpoint health

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379,http://localhost:42379 \
  member list --write-out=table

ETCDCTL_API=3 /tmp/etcd-download-test/etcdctl \
  --endpoints=http://localhost:12379,http://localhost:22379,http://localhost:32379,http://localhost:42379 \
  endpoint status --write-out=table
##################################
```

**Comment by lyddragon** [comment]
my cluster have 3 member, and do snapshot from 3 member cluster!!!

**Comment by gyuho** [comment]
> do snapshot from 3 member cluster!!!

what do you mean?

**Comment by lyddragon** [comment]
Follow my steps.

**Comment by lyddragon** [comment]
the very beginning，my cluster have 3 member，your cluster only one member.

**Comment by lyddragon** [comment]
set up 3-node cluster (A, B, C)
get snapshot of A
shut down A B C
restore A B C snapshot
remove db directory
restart A B C with same configuration 
add member D
start member D
get keys on member D

1 2 3 4 5 6 7 8 9

**Comment by lyddragon** [comment]
sorry
set up 3-node cluster (A, B, C)
get snapshot of A
shut down A B C
remove A B C db directory
restore A B C snapshot
restart A B C with same configuration
add member D
start member D
get keys on member D

1 2 3 4 5 6 7 8 9

**Comment by gyuho** [comment]
Updated https://github.com/coreos/etcd/issues/9109#issuecomment-356150234. Still works, or I am still missing something. As long as the snapshot from A is recent (contains all keys), it won't make any difference. Please provide locally reproducible script as above, or try that yourself.

**Comment by lyddragon** [comment]
associated with it: wal-dir
my configuration:
name: "nodename"
data-dir: /tmp/nodename.data
wal-dir: /tmp/nodename.wal.data
......

**Comment by gyuho** [comment]
Now I see what went wrong. You are configuring separate WAL directory `/data/etcd/intranet-test.wal.data` when `snapshot restore` outputs everything under `--data-dir`. Please change your `--wal-dir` to `filepath.Join(DATA_DIR, "member", "wal")`.

**Comment by lyddragon** [comment]
--wal-dir
Path to the dedicated wal directory. If this flag is set, etcd will write the WAL files to the walDir rather than the dataDir. This allows a dedicated disk to be used, and helps avoid io competition between logging and other IO operations.
default: ""
env variable: ETCD_WAL_DIR

**Comment by gyuho** [comment]
Problem is `snapshot restore` command outputs everything under `--data-dir` flag. And as you said

> but 'etcdctl snapshot restore ' no option: --wal-dir

So, I assume your systemd config points to wrong wal dir.

**Comment by lyddragon** [comment]
'etcdctl snapshot restore' need to add option: --wal-dir, or add option: --config-file (may be necessary) .
--initial-cluster-state  option should be abandoned.

**Comment by gyuho** [comment]
@lyddragon Can you open a new issue explaining your use case? Thanks!

**Comment by lyddragon** [comment]
To make it easier to maintain

**Comment by gyuho** [comment]
@lyddragon I submitted a PR to add `snapshot restore --wal-dir` flag. Will backport to v3.3.

**Comment by lyddragon** [comment]
If add --config-file may be more convenient.
